tiledb/sm/c_api/tiledb.h

```diff
@@ -2725,6 +2725,41 @@ TILEDB_EXPORT int32_t tiledb_array_alloc(
 TILEDB_EXPORT int32_t tiledb_array_open(
     tiledb_ctx_t* ctx, tiledb_array_t* array, tiledb_query_type_t query_type);
 
+/**
+ * Similar to `tiledb_array_open`, but this function takes as input a
+ * timestamp, representing time in milliseconds ellapsed since
+ * 1970-01-01 00:00:00 +0000 (UTC). Opening the array at a
+ * timestamp provides a view of the array with all writes/updates that
+ * happened at or before `timestamp` (i.e., excluding those that
+ * occurred after `timestamp`). This function is useful to ensure
+ * consistency at a potential distributed setting, where machines
+ * need to operate on the same view of the array.
+ *
+ * **Example:**
+ *
+ * @code{.c}
+ * tiledb_array_t* array;
+ * tiledb_array_alloc(ctx, "hdfs:///tiledb_arrays/my_array", &array);
+ * // Assuming `timestamp` is time represented in milliseconds:
+ * tiledb_array_open_at(ctx, array, TILEDB_READ, timestamp);
+ * @endcode
+ *
+ * @param ctx The TileDB context.
+ * @param array The array object to be opened.
+ * @param query_type The type of queries the array object will be receiving.
+ * @param timestamp The timestamp to open the array at.
+ * @return `TILEDB_OK` for success and `TILEDB_ERR` for error.
+ *
+ * @note If the same array object is opened again without being closed,
+ *     an error will be thrown.
+ * @note This function is applicable only to read queries.
+ */
+TILEDB_EXPORT int32_t tiledb_array_open_at(
```

Comment:
Would it be useful to add a `uint64_t TILEDB_TIMESTAMP_ALL` constant value? Then you could use `open_at` even when you want the latest timestamp. Just thinking this is similar to allowing open_with_key to use a null key for unencrypted arrays.

Reply:
Do you perhaps mean `TILEDB_TIMESTAMP_MAX`? But then you can always use `UINT64_MAX`. Do you think it would be valuable to use another constant for that?

What I do find valuable is exposing a macro for returning the current timestamp with `tiledb::sm::utils::time::timestamp_ms()`.